File: apps/java/src/main/java/chapter09/DelimitedFiles.java
package chapter09;

import chapter02.Row;
import chapter02.Table;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/** 文字コードと区切り文字を指定して、区切り文字で区切ったファイルを {@link Table} に読み込む。 */
public final class DelimitedFiles {
  private DelimitedFiles() {}

  /** 1 行目を列名として読み込む。文字コードが違えば MalformedInputException を投げる。 */
  public static Table load(Path file, Charset charset, String delimiter) throws IOException {
    String separator = Pattern.quote(delimiter);
    List<String> lines = Files.readAllLines(file, charset);
    List<String> columns = List.of(lines.getFirst().split(separator));
    List<Row> rows =
        lines.stream()
            .skip(1)
            .filter(line -> !line.isBlank())
            .map(line -> toRow(columns, line.split(separator, -1)))
            .toList();
    return new Table(columns, rows);
  }

  private static Row toRow(List<String> columns, String[] values) {
    Map<String, String> cells = new HashMap<>();
    for (int i = 0; i < columns.size(); i++) {
      cells.put(columns.get(i), values[i]);
    }
    return new Row(cells);
  }
}

File: apps/scala/src/main/scala/machinelearning/chapter09/DelimitedFiles.scala
package machinelearning.chapter09

import java.nio.charset.Charset
import java.nio.file.{Files, Path}
import java.util.regex.Pattern
import machinelearning.chapter02.{Row, Table}
import scala.jdk.CollectionConverters.*

/** 文字コードと区切り文字を指定して、区切り文字で区切ったファイルを [[machinelearning.chapter02.Table]] に読み込む。 */
object DelimitedFiles:

  /** 1 行目を列名として読み込む。文字コードが違えば MalformedInputException を投げる。 */
  def load(file: Path, charset: Charset, delimiter: String): Table =
    val separator = Pattern.quote(delimiter)
    val lines = Files.readAllLines(file, charset).asScala.toVector
    val columns = lines.head.split(separator, -1).toVector
    val rows = lines.tail.filter(_.trim.nonEmpty).map(toRow(columns, _, separator))
    Table(columns, rows)

  private def toRow(columns: Vector[String], line: String, separator: String): Row =
    Row(columns.zip(line.split(separator, -1).toVector).toMap)
